### PR TITLE
Homework - cars

### PR DESCRIPTION
**Pull request by mateusz-talma**
1. Added InvalidGear exception
2. Added InvalidSpeed exception
3. Added InvalidChangeEngine exception
4. Added InvalidCharge exception
5. Encapsulate data (protected, private section)
6. Added setters and getters for ElectricCars
7. Delete all new allocation. I used unique_ptr and check programs with valgrinds - no leaks memory
8. Added ChangeEngine function with swap engine - for details see test cases
9. Some validations (speed, electric car charging, gearbox change etc.)
10. Added tests
11. Added workflow

### CODE REVIEW

**Review by WBigaj — COMMENTED** [review]
I have checked your code and it looks good :)

**Review by MaciejSalamonski — APPROVED** [review]
Nice code! i dont have any comments. Have a great daty!

**Review by Kasperice — APPROVED** [review]
Good code, I have no comments :)

File: ElectricCar.cpp [hunk 1]
```diff
@@ -1,16 +1,40 @@
...
+
+std::unique_ptr<ElectricEngine> ElectricCar::changeEngine(std::unique_ptr<ElectricEngine> newEngine) {
+    if (speed_ != 0) {
+        throw InvalidChangeEngine("You have to stop the car if you would like change engine");
+    }
+
+    std::unique_ptr<ElectricEngine> tmp = std::move(newEngine);
+    std::swap(tmp, electricEngine_);
+    return tmp;
+}
```
Comment (ziobron) [inline ElectricCar.cpp:40]: nice implementation!

File: PetrolCar-ut.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+TEST_F(PetrolCarTest, ShouldNotChangeEngineIfCarIsNotStopped) {
+    petrolCar.accelerate(100);
+    EXPECT_THROW(petrolCar.changeEngine(std::make_unique<PetrolEngine>(100, 100, 5)), InvalidChangeEngine);
+}
+
+TEST_F(PetrolCarTest, ChangeEngineShouldSwapEngines) {
+    petrolCar.accelerate(0);
+    auto newEngine = std::make_unique<PetrolEngine>(20, 20, 1);     // one gear
```
Comment (ziobron) [inline PetrolCar-ut.cpp:55]: I can see a few magic values here. They are not important from test POV.

File: PetrolEngine.hpp [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+enum class GearBox {
+    rear = -1,
+    neutral = 0,
+    first = 1,
+    second = 2,
+    third = 3,
+    fourth = 4,
+    fifth = 5,
+    sixth = 6
+};
```
Comment (ziobron) [inline PetrolEngine.hpp:12]: this is great 👍

**Review by ziobron — APPROVED** [review]
Nice idea with Gearbox. Really good solution :)

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
9 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
5 XP za wyróżnione rozwiązanie (1120)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119, 1120] }

**Pull request closed without merging**